### PR TITLE
17 add ngx translate conversion

### LINKED ISSUE #17: Add `ngx-translate` conversation to `pkg/convert/json_ngxtranslate.go`
`pkg/convert` contains all conversations between different translation files and a translate data model.

Convert to and from Angular ngx-translate.

Example of Angular JSON. `ngx-translate`. The lookup can be nested or dot concatenated string (e.g. `home.title`).

```json
{
  "home": {
    "title": "Welcome!"
  }
}
```

The target is serialized []byte.

### PR DESCRIPTION
**Pull request by VladislavsPerkanuks**
Output is nested JSON, don't remember if we wanted output as flat or nested, but nested seems simpler

### CODE REVIEW

File: pkg/convert/json_ngxtranslate.go [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+	"encoding/json"
+	"errors"
+	"fmt"
+
+	"github.com/expect-digital/translate/pkg/model"
+)
+
+var ErrWrongKey = errors.New("Wrong key type")
+
+func fromNgxTranslate(b []byte) (messages model.Messages, err error) {
```
Comment (DarjaLeonova) [inline pkg/convert/json_ngxtranslate.go]: Shouldn't we use first letter upper-case if we want use these functions outside this file?

File: pkg/convert/json_ngxtranslate.go [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+	}
+
+	if err = traverseMap("", dst); err != nil {
+		return messages, fmt.Errorf("traverse ngx: %w", err)
+	}
+
+	return messages, nil
+}
+
+func toNgxTranslate(messages model.Messages) (b []byte, err error) {
```
Comment (DarjaLeonova) [inline pkg/convert/json_ngxtranslate.go]: Same question about this function

File: pkg/convert/json_ngxtranslate_test.go [hunk 3]
```diff
@@ -0,0 +1,129 @@
+package convert
+
+import (
+	"testing"
+
+	"github.com/expect-digital/translate/pkg/model"
+	"github.com/stretchr/testify/assert"
+)
+
+func Test_fromNgxTranslate(t *testing.T) {
```
Comment (DarjaLeonova) [inline pkg/convert/json_ngxtranslate_test.go:10]: I'm not sure, but should we commit tests also? I remember that @janishorsts said that its not needed

File: pkg/convert/json_ngxtranslate.go [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+	"errors"
+	"fmt"
+
+	"github.com/expect-digital/translate/pkg/model"
+)
+
+var ErrWrongKey = errors.New("Wrong key type")
+
+func fromNgxTranslate(b []byte) (messages model.Messages, err error) {
+	var dst map[string]interface{}
```
Comment (DarjaLeonova) [inline pkg/convert/json_ngxtranslate.go:12]: What means dst variable? Just asking ;D

**Review by DarjaLeonova — CHANGES_REQUESTED** [review]

File: go.mod [hunk 5]
```diff
@@ -3,18 +3,23 @@ module github.com/expect-digital/translate
...
-	google.golang.org/genproto v0.0.0-20200526211855-cb27e3aa2013 // indirect
 	gopkg.in/yaml.v3 v3.0.1 // indirect
 )
+
+require (
+	github.com/golang/protobuf v1.5.2 // indirect
+	github.com/stretchr/testify v1.8.1
+	golang.org/x/net v0.3.0 // indirect
+	golang.org/x/sys v0.3.0 // indirect
+)
```
Comment (janishorsts) [inline go.mod]: Merge all indirect dependencies under one `require`.

File: pkg/convert/json_ngxtranslate.go (hunk 4 above)
Comment (janishorsts) [reply pkg/convert/json_ngxtranslate.go:12]: `dst` is short for `destination`

File: pkg/convert/json_ngxtranslate.go [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+
+	if err = json.Unmarshal(b, &dst); err != nil {
+		return messages, fmt.Errorf("unmarshal ngx translate: %w", err)
+	}
+
+	traverseMap := func(key string, value any) (err error) {
```
Comment (janishorsts) [inline pkg/convert/json_ngxtranslate.go]: Declare only

```suggestion
	var traverseMap func(key string, value any) error
```

File: pkg/convert/json_ngxtranslate.go [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+		return err
+	}
+
+	traverseMap = func(key string, v any) (err error) {
+		switch value := v.(type) {
```
Comment (janishorsts) [inline pkg/convert/json_ngxtranslate.go]: Advisory. I'd swap names for consistency.

```suggestion
	traverseMap = func(key string, value interface{}) (err error) {
		switch v := value.(type) {
```

File: pkg/convert/json_ngxtranslate.go [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+		case string:
+			messages.Messages = append(messages.Messages, model.Message{ID: key, Message: value})
+		case map[string]interface{}:
+			for key2, val2 := range value {
+				var newKey string
+				if key != "" {
+					newKey += key + "."
+				}
+
+				newKey += key2
```
Comment (janishorsts) [inline pkg/convert/json_ngxtranslate.go]: Keep it simple, you can still keep it clear and nice without declaring a new variable.

Alternative names `subKey`, `subValue` to `key2`, `value2`.

```suggestion
				if key != "" {
					subKey = key + "." + subKey
				}
```

File: pkg/convert/json_ngxtranslate.go [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+
+	"github.com/expect-digital/translate/pkg/model"
+)
+
+var ErrWrongKey = errors.New("Wrong key type")
```
Comment (janishorsts) [inline pkg/convert/json_ngxtranslate.go]: 1) In go, it's idiomatic to start an error message with a lowercase letter.
2) Declare package errors only if the package consumers require to check for specific errors. I'm not sure if this is applicable to the given package.

```suggestion
var ErrWrongKey = errors.New("wrong key type")
```

File: pkg/convert/json_ngxtranslate_test.go [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+						ID:      "hello",
+						Message: "world",
+					},
+					{
+						ID:      "hello.beautiful",
+						Message: "world",
+					},
+				},
+			},
+			wantB:   []byte(`[{"ID":"hello","Message":"world"},{"ID":"hello.beautiful","Message":"world"}]`),
```
Comment (janishorsts) [inline pkg/convert/json_ngxtranslate_test.go]: Incorrect expectation in the test

```suggestion
			wantB:   []byte(`{"hello":"world","hello.beautiful":"world"}`),
```

**Review by janishorsts — CHANGES_REQUESTED** [review]

File: pkg/convert/json_ngxtranslate_test.go (hunk 3 above)
Comment (VladislavsPerkanuks) [reply pkg/convert/json_ngxtranslate_test.go:10]: It was complicated function with recursion and I needed to make sure that it does the thing I want it to do. So I made simple tests

**Review by janishorsts — APPROVED** [review]

**Pull request merged**